# DP: On linux targets pass --as-needed by default to the linker.

--- ./src/gcc/config/rs6000/sysv4.h~	2010-11-12 18:51:36.221896517 +0100
+++ ./src/gcc/config/rs6000/sysv4.h	2010-11-12 18:57:58.751894768 +0100
@@ -914,7 +914,7 @@
 
 #if defined(HAVE_LD_EH_FRAME_HDR)
 # undef LINK_EH_SPEC
-# define LINK_EH_SPEC "--no-add-needed %{!static:--eh-frame-hdr} "
+# define LINK_EH_SPEC "--no-add-needed --as-needed %{!static:--eh-frame-hdr} "
 #endif
 
 #define CPP_OS_LINUX_SPEC "-D__unix__ -D__gnu_linux__ -D__linux__ \
--- ./src/gcc/config/alpha/elf.h~	2010-11-12 18:51:36.221896517 +0100
+++ ./src/gcc/config/alpha/elf.h	2010-11-12 18:57:49.309394525 +0100
@@ -441,7 +441,7 @@
    I imagine that other systems will catch up.  In the meantime, it
    doesn't harm to make sure that the data exists to be used later.  */
 #if defined(HAVE_LD_EH_FRAME_HDR)
-#define LINK_EH_SPEC "--no-add-needed %{!static:--eh-frame-hdr} "
+#define LINK_EH_SPEC "--no-add-needed --as-needed %{!static:--eh-frame-hdr} "
 #endif
 
 /* A C statement (sans semicolon) to output to the stdio stream STREAM
--- ./src/gcc/config/ia64/linux.h~	2010-11-12 18:51:36.221896517 +0100
+++ ./src/gcc/config/ia64/linux.h	2010-11-12 18:57:37.399395163 +0100
@@ -82,7 +82,7 @@
    Signalize that because we have fde-glibc, we don't need all C shared libs
    linked against -lgcc_s.  */
 #undef LINK_EH_SPEC
-#define LINK_EH_SPEC "--no-add-needed "
+#define LINK_EH_SPEC "--no-add-needed --as-needed "
 
 #define MD_UNWIND_SUPPORT "config/ia64/linux-unwind.h"
 
--- ./src/gcc/config/linux.h~	2010-11-12 18:51:36.221896517 +0100
+++ ./src/gcc/config/linux.h	2010-11-12 18:57:14.409395122 +0100
@@ -89,7 +89,7 @@
     } while (0)
 
 #if defined(HAVE_LD_EH_FRAME_HDR)
-#define LINK_EH_SPEC "--no-add-needed %{!static:--eh-frame-hdr} "
+#define LINK_EH_SPEC "--no-add-needed --as-needed %{!static:--eh-frame-hdr} "
 #endif
 
 /* Define this so we can compile MS code for use with WINE.  */
